### PR TITLE
QCQMC Part 7: Add Experiment classes

### PR DESCRIPTION
**Pull request by fdmalone**
Depends upon #351 and #353: Adds the Experiment class which packages everything together to run a simulated experiment.

### CODE REVIEW

File: recirq/qcqmc/experiment.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import attrs
+import cirq
+import numpy as np
+import qsimcirq
+from pytz import timezone
+
+from recirq.qcqmc import blueprint, config, data, for_refactor
+
+
+def _to_tuple(x: Optional[Iterable[float]]) -> Optional[Sequence[float]]:
```
Comment (dstrain115) [inline recirq/qcqmc/experiment.py:27]: I would consolidate this with other tuple functions or remove it.  This is simple enough we could add it inline, 
such as `tuple(x) if x is None else x`

File: recirq/qcqmc/experiment.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    raw_samples: np.ndarray = attrs.field(converter=np.asarray)
+    metadata: Dict[str, Any] = attrs.field(factory=dict)
+
+    def _json_dict_(self):
+        simple_dict = attrs.asdict(self)
+        simple_dict["params"] = self.params
+        return simple_dict
+
+
+def build_experiment(
```
Comment (dstrain115) [inline recirq/qcqmc/experiment.py:94]: Same here.  Add as a classmethod on experiment data?

File: recirq/qcqmc/experiment_test.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import cirq
+
+from recirq.qcqmc.blueprint import BlueprintData
+from recirq.qcqmc.experiment import SimulatedExperimentParams, build_experiment
+from recirq.qcqmc.hamiltonian import HamiltonianData
+from recirq.qcqmc.trial_wf import TrialWavefunctionData
+
+
+def test_small_experiment_raw_samples_shape(
```
Comment (dstrain115) [inline recirq/qcqmc/experiment_test.py:25]: This test file seems short.  Does it test everything in experiments?

File: recirq/qcqmc/experiment.py (hunk 1 above)
Comment (fdmalone) [reply recirq/qcqmc/experiment.py:27]: Done.

File: recirq/qcqmc/experiment.py (hunk 2 above)
Comment (fdmalone) [reply recirq/qcqmc/experiment.py:94]: Done.

File: recirq/qcqmc/experiment_test.py (hunk 3 above)
Comment (fdmalone) [reply recirq/qcqmc/experiment_test.py:25]: It does at least call every function (after a small update) but the test coverage is not great. I think I'd @wjhuggins to help writing some sensible physical tests of the shadow tomography protocol but I think that can be a follow up (we discussed writing some better documentation of the shadow tomography part of the code which is probably of most interest to QIS people).

Comment (dstrain115) [reply recirq/qcqmc/experiment_test.py:25]: Ok, yeah, let's make it a follow up PR.

**Review by dstrain115 — APPROVED** [review]

**Pull request merged**